src/drivers/distance_sensor/tfmini/tfmini.cpp

```diff
@@ -534,43 +531,56 @@ TFMINI::collect()
 	char readbuf[sizeof(_linebuf)];
 	unsigned readlen = sizeof(readbuf) - 1;
 
-	/* read from the sensor (uart buffer) */
-	ret = ::read(_fd, &readbuf[0], readlen);
-
-	if (ret < 0) {
-		DEVICE_DEBUG("read err: %d", ret);
-		perf_count(_comms_errors);
-		perf_end(_sample_perf);
-
-		/* only throw an error if we time out */
-		if (read_elapsed > (_conversion_interval * 2)) {
-			return ret;
+	int ret = 0;
+	float distance_m = -1.0f;
 
-		} else {
-			return -EAGAIN;
-		}
+	/* Check the number of bytes available in the buffer*/
+	int bytes_available = 0;
+	::ioctl(_fd, FIONREAD, (unsigned long)&bytes_available);
 
-	} else if (ret == 0) {
+	if (!bytes_available) {
 		return -EAGAIN;
 	}
 
-	_last_read = hrt_absolute_time();
+	/* parse entire buffer */
+	do {
+		/* read from the sensor (uart buffer) */
+		ret = ::read(_fd, &readbuf[0], readlen);
+
+		if (ret < 0) {
+			PX4_ERR("read err: %d", ret);
+			perf_count(_comms_errors);
+			perf_end(_sample_perf);
+
+			/* only throw an error if we time out */
+			if (read_elapsed > (_conversion_interval * 2)) {
+				/* flush anything in RX buffer */
+				tcflush(_fd, TCIFLUSH);
+				return ret;
+
+			} else {
+				return -EAGAIN;
+			}
+		}
 
-	float distance_m = -1.0f;
-	bool valid = false;
+		_last_read = hrt_absolute_time();
 
-	for (int i = 0; i < ret; i++) {
-		if (OK == tfmini_parser(readbuf[i], _linebuf, &_linebuf_index, &_parse_state, &distance_m)) {
-			valid = true;
+		/* parse buffer */
+		for (int i = 0; i < ret; i++) {
+			tfmini_parse(readbuf[i], _linebuf, &_linebuf_index, &_parse_state, &distance_m);
 		}
-	}
 
-	if (!valid) {
+		/* bytes left to parse */
+		bytes_available -= ret;
```

Comment:
If for some reason `read` returns 0, you'll have an infinite loop.

Reply:
Thanks for the catch. Will fix.